From 7946975419366eddda41d1bb4bade6b956b52b1b Mon Sep 17 00:00:00 2001
From: Gaspard Reulet <gaspard.reulet@usherbrooke.ca>
Date: Mon, 9 Mar 2026 09:29:52 -0400
Subject: [PATCH] fix: change lookup return value when one col selected

Fix #4001
Lookup function can be used to retrieve column(s) in a pandas df.
For a given query, lookup returns a list of results.
This behaviour is counterintuitive when 1 col is given `cols="mycol"`.

Change lookup function return value when `cols="mycol"` to `value`,
instead of `list(value)`.
Return None if item is "None-like", eg: pd.NA, np.nan.

fix: remove none-like check and return as is

fix: add test case and update docs

Clarify behavior of single item return in lookup

Update documentation for lookup function

Clarified return value behaviour for single column lookups.
---
 docs/snakefiles/rules.rst       |  3 +++
 src/snakemake/ioutils/lookup.py | 11 ++++++++---
 tests/test_ioutils/Snakefile    |  1 +
 3 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/docs/snakefiles/rules.rst b/docs/snakefiles/rules.rst
index db3f08f7..9ed9ab32 100644
--- a/docs/snakefiles/rules.rst
+++ b/docs/snakefiles/rules.rst
@@ -457,6 +457,8 @@ If the query results in multiple rows, the result is returned as a list of
 named tuples with the column names as attributes.
 If the query results in a single row, the result is returned as a single
 named tuple with the column names as attributes.
+If the query results in a single row and the cols parameter is a string,
+the result is returned as a single item.
 If the query or dpath parameter is given a function, the function will be evaluated with wildcards passed as the first argument.
 In case of dpath, if the dpath is not found, a ``LookupError`` is raised, unless a
 default fallback value is provided via the ``default`` argument (this argument is ignored in case of ``query``).
@@ -500,6 +502,7 @@ or to a single column, e.g.
     lookup(query="sample == '{sample}'", within=samples, cols="somecolumn")
 
 In the latter case, just a list of items in that column is returned (e.g. ``["a", "b", "c"]``).
+Also in the latter case, if the query results in only one row, then the single item is returned (e.g. ``"a"``).
 
 The argument ``is_nrows`` allows to test for a given number of rows in the queried dataframe.
 If it is used, lookup just returns a boolean value indicating whether the number of rows in the queried dataframe matches the given number:
diff --git a/src/snakemake/ioutils/lookup.py b/src/snakemake/ioutils/lookup.py
index f19a6433..6ce353c6 100644
--- a/src/snakemake/ioutils/lookup.py
+++ b/src/snakemake/ioutils/lookup.py
@@ -1,7 +1,7 @@
+import re
 from abc import ABC, abstractmethod
 from collections.abc import Mapping
 from functools import partial
-import re
 from typing import List, Optional, Union
 
 import snakemake.io
@@ -135,7 +135,8 @@ def lookup(
     ``lookup(query="sample == '{sample}'", within=samples, cols=["somecolumn"])`` or to
     a single column, e.g.
     ``lookup(query="sample == '{sample}'", within=samples, cols="somecolumn")``.
-    In the latter case, just a list of items in that column is returned.
+    In the latter case, the scalar value is returned when the query matches a
+    single row, or a list of values when it matches multiple rows.
     Finally, if the integer argument ``is_nrows`` is used, this returns true
     if there are that many rows in the query results, false otherwise.
 
@@ -192,7 +193,11 @@ def lookup(
                 res = res[cols]
                 if not isinstance(cols, list):
                     # single column select, just return a list of values
-                    return res.to_list()
+                    res = res.to_list()
+                    if len(res) == 1:
+                        # just return the value if it is only one
+                        return res[0]
+                    return res
             res = list(res.itertuples(index=cols is None))
             if len(res) == 1:
                 # just return the item if it is only one
diff --git a/tests/test_ioutils/Snakefile b/tests/test_ioutils/Snakefile
index 5061e06e..127e23b0 100644
--- a/tests/test_ioutils/Snakefile
+++ b/tests/test_ioutils/Snakefile
@@ -8,6 +8,7 @@ configfile: "config.yaml"
 
 assert lookup(dpath="does/not/exist", within=config, default=None) is None
 assert lookup(dpath="does/not/exist", within=config, default=5) == 5
+assert lookup(query="sample == 1", within=samples, cols="val") == 1
 assert (
     extract_checksum("samples.md5", file="1.tsv") == "9695eb6f38992d796551f4cb20d7d138"
 )
-- 
2.52.0

